File: extensions/path.py
import os
import json
import debugutils
from wutilerror import check_types
mark = debugutils.mark_time
log = debugutils.log

CONFIG_DIR = os.path.join(os.path.expanduser("~"), ".wutil")
PATH_FILE = os.path.join(CONFIG_DIR, "paths.json")
REAL_CWD = os.getenv("WUTIL_REAL_CWD")
def load_paths():
    if not os.path.exists(PATH_FILE):
        log("paths.json does not exist yet", important=False, source="path")
        return {}
    try:
        with open(PATH_FILE, "r", encoding="utf-8") as f:
            log("loading alias paths from disk", important=False, source="path")
            return json.load(f)
    except:
        log("failed to load alias paths; returning empty set", source="path")
        return {}

def save_paths(data):
    os.makedirs(CONFIG_DIR, exist_ok=True)
    log(f"saving {len(data)} aliases", important=False, source="path")
    with open(PATH_FILE, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=4)


class Extension:
    def __init__(self):
        self.name = "path"
        self.desc = "Manage WUtil executable aliases."
        self.args = ["action", "arg1", "arg2"]
        self.short = "p"
        self.requires_window = False

    def main(self, action=None, arg1=None, arg2=None):
        mark("path start", source="path")
        log(f"path action={action} arg1={arg1} arg2={arg2}", important=False, source="path")
        check_types(
            action=(action, str, False),
            arg1=(arg1, str, True),
            arg2=(arg2, str, True)
        )

        if action == "add":
            return self._add(arg1, arg2)
        elif action == "delete":
            return self._delete(arg1)
        elif action == "list":
            return self._list()
        else:
            print("Usage:")
            print("  wutil path add <file> <alias>")
            print("  wutil path delete <alias>")
            print("  wutil path list")

    def _add(self, file, alias):
        if not file or not alias:
            print("Usage: wutil path add <file> <alias>")
            log("path add aborted because file or alias was missing", source="path")
            return

        # Convert relative → absolute, but from the user's real cwd
        file = os.path.abspath(os.path.join(REAL_CWD, file))
        log(f"resolved alias target path={file}", important=False, source="path")

        if not os.path.exists(file):
            print(f"Error: file not found: {file}")
            log(f"path add failed because target was missing: {file}", source="path")
            return

        paths = load_paths()

        if alias in paths:
            print(f"Alias '{alias}' already exists.")
            log(f"path add rejected duplicate alias={alias}", source="path")
            return

        paths[alias] = file
        save_paths(paths)
        print(f"Added alias '{alias}' → {file}")
        log(f"alias added alias={alias}", source="path")

    def _delete(self, alias):
        if not alias:
            print("Usage: wutil path delete <alias>")
            log("path delete aborted because alias was missing", source="path")
            return

        paths = load_paths()

        if alias not in paths:
            print(f"Alias '{alias}' does not exist.")
            log(f"path delete could not find alias={alias}", source="path")
            return

        removed = paths.pop(alias)
        save_paths(paths)
        print(f"Deleted alias '{alias}' (was → {removed})")
        log(f"alias deleted alias={alias}", source="path")

    def _list(self):
        paths = load_paths()

        if not paths:
            print("No aliases stored.")
            log("path list found no aliases", source="path")
            return

        for alias, path in paths.items():
            print(f"{alias:<12} → {path}")
        log(f"listed {len(paths)} aliases", source="path")

File: extensions/path.go
package main

//wutil:name path
//wutil:short p
//wutil:desc Manage WUtil executable aliases.
//wutil:args action,arg1,arg2
//wutil:requires_window false

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"

	"windowutil/pkg/wutilext"
)

func main() {
	action, arg1, arg2 := "", "", ""
	if len(os.Args) > 1 {
		action = os.Args[1]
	}
	if len(os.Args) > 2 {
		arg1 = os.Args[2]
	}
	if len(os.Args) > 3 {
		arg2 = os.Args[3]
	}

	wutilext.Mark("path start", "path")
	wutilext.LogDetail(fmt.Sprintf("path action=%s arg1=%s arg2=%s", action, arg1, arg2), "path")

	switch action {
	case "add":
		add(arg1, arg2)
	case "delete":
		del(arg1)
	case "list":
		list()
	default:
		fmt.Println("Usage:")
		fmt.Println("  wutil path add <file> <alias>")
		fmt.Println("  wutil path delete <alias>")
		fmt.Println("  wutil path list")
	}
}

func add(file string, alias string) {
	if file == "" || alias == "" {
		fmt.Println("Usage: wutil path add <file> <alias>")
		wutilext.Log("path add aborted because file or alias was missing", "path")
		return
	}
	if !filepath.IsAbs(file) {
		file = filepath.Join(wutilext.RealCWD(), file)
	}
	file, _ = filepath.Abs(file)
	wutilext.LogDetail(fmt.Sprintf("resolved alias target path=%s", file), "path")
	if _, err := os.Stat(file); err != nil {
		fmt.Printf("Error: file not found: %s\n", file)
		wutilext.Log(fmt.Sprintf("path add failed because target was missing: %s", file), "path")
		return
	}

	paths := loadPaths()
	if _, exists := paths[alias]; exists {
		fmt.Printf("Alias '%s' already exists.\n", alias)
		wutilext.Log(fmt.Sprintf("path add rejected duplicate alias=%s", alias), "path")
		return
	}
	paths[alias] = file
	savePaths(paths)
	fmt.Printf("Added alias '%s' -> %s\n", alias, file)
	wutilext.Log(fmt.Sprintf("alias added alias=%s", alias), "path")
}

func del(alias string) {
	if alias == "" {
		fmt.Println("Usage: wutil path delete <alias>")
		wutilext.Log("path delete aborted because alias was missing", "path")
		return
	}
	paths := loadPaths()
	removed, exists := paths[alias]
	if !exists {
		fmt.Printf("Alias '%s' does not exist.\n", alias)
		wutilext.Log(fmt.Sprintf("path delete could not find alias=%s", alias), "path")
		return
	}
	delete(paths, alias)
	savePaths(paths)
	fmt.Printf("Deleted alias '%s' (was -> %s)\n", alias, removed)
	wutilext.Log(fmt.Sprintf("alias deleted alias=%s", alias), "path")
}

func list() {
	paths := loadPaths()
	if len(paths) == 0 {
		fmt.Println("No aliases stored.")
		wutilext.Log("path list found no aliases", "path")
		return
	}
	aliases := make([]string, 0, len(paths))
	for alias := range paths {
		aliases = append(aliases, alias)
	}
	sort.Strings(aliases)
	for _, alias := range aliases {
		fmt.Printf("%-12s -> %s\n", alias, paths[alias])
	}
	wutilext.Log(fmt.Sprintf("listed %d aliases", len(paths)), "path")
}

func loadPaths() map[string]string {
	data, err := os.ReadFile(pathFile())
	if err != nil {
		wutilext.LogDetail("paths.json does not exist yet", "path")
		return map[string]string{}
	}
	paths := map[string]string{}
	if err := json.Unmarshal(data, &paths); err != nil {
		wutilext.Log("failed to load alias paths; returning empty set", "path")
		return map[string]string{}
	}
	wutilext.LogDetail("loading alias paths from disk", "path")
	return paths
}

func savePaths(paths map[string]string) {
	file := pathFile()
	_ = os.MkdirAll(filepath.Dir(file), 0755)
	data, _ := json.MarshalIndent(paths, "", "    ")
	_ = os.WriteFile(file, data, 0644)
	wutilext.LogDetail(fmt.Sprintf("saving %d aliases", len(paths)), "path")
}

func pathFile() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".wutil", "paths.json")
}
